From 7263023efd095fd8fe2afd7d53efc915b60d4cfa Mon Sep 17 00:00:00 2001
From: wanlong <long.wan@spacemit.com>
Date: Sat, 12 Oct 2024 15:08:38 +0800
Subject: [PATCH 1174/1448] wireless: rtl8852bs: fix oops in
 rtw_get_chplan_hdl()

Unable to handle kernel paging request at virtual address ffffffc816103c40
Oops [#1]
CPU: 1 PID: 1039 Comm: disp_eng_share_ Tainted: G        W          6.6.36 #2.0~rc7.2
Hardware name: spacemit k1-x MUSE-Pi board (DT)
epc : rtw_get_chplan_hdl+0x1da/0x238 [8852bs]
 ra : rtw_get_chplan_hdl+0x1d6/0x238 [8852bs]
epc : ffffffff022a9bc0 ra : ffffffff022a9bbc sp : ffffffc802073c30
 gp : ffffffff820e6640 tp : ffffffd91a44b300 t0 : 0000000000000000
 t1 : ffffffd90bf0a6a8 t2 : 0000000000000000 s0 : ffffffc802073ce0
 s1 : ffffffc80d2a5000 a0 : ffffffc80d2a5078 a1 : ffffffd9113fea50
 a2 : 0000000000000060 a3 : ffffffd9113fea50 a4 : ffffffffffffffd9
 a5 : ffffffc816103c40 a6 : 0000000000000000 a7 : 0000000000000000
 s2 : ffffffd9113fe000 s3 : ffffffd9113ff000 s4 : ffffffc802073c68
 s5 : ffffffc80d2a5028 s6 : ffffffc80d2a5018 s7 : ffffffc802073c4c
 s8 : ffffffd91141b200 s9 : 0000000000000003 s10: ffffffd9113ff000
 s11: ffffffc80d2a5078 t3 : 0000000000000000 t4 : 0000000000000000
 t5 : ffffffd90bf0a668 t6 : ffffffc80d2a5858
status: 0000000200000120 badaddr: ffffffc816103c40 cause: 000000000000000f
[<ffffffff022a9bc0>] rtw_get_chplan_hdl+0x1da/0x238 [8852bs]
[<ffffffff0225db1e>] rtw_run_cmd+0x88/0x252 [8852bs]
[<ffffffff022d9c2e>] phl_run_core_cmd+0x4e/0x1b0 [8852bs]
[<ffffffff0233fdbe>] _phl_cmd_complete+0x168/0x320 [8852bs]
[<ffffffff02337884>] push_back_idle_msg+0x48/0x136 [8852bs]
[<ffffffff02338d8a>] msg_dispatch+0x2c0/0x3e2 [8852bs]
[<ffffffff0233c45e>] dispr_thread_loop_hdl+0x25e/0x2d4 [8852bs]
[<ffffffff0233c4e4>] dispr_share_thread_loop_hdl+0x10/0x18 [8852bs]
[<ffffffff0233607c>] share_thread_hdl+0xa0/0x194 [8852bs]
[<ffffffff8003ef16>] kthread+0xda/0xf6
[<ffffffff80f24a16>] ret_from_fork+0xe/0x1c
Code: 0593 2709 8513 0784 5097 fff8 80e7 7000 3783 000c (e384) 4501
---[ end trace 0000000000000000 ]---

Change-Id: I96f7da7edeb0fb74e3e244460e5f7e96b3a98efc
---
 .../net/wireless/realtek/rtl8852bs/core/rtw_chplan.c | 12 +++++++-----
 .../net/wireless/realtek/rtl8852bs/core/rtw_chplan.h |  2 +-
 2 files changed, 8 insertions(+), 6 deletions(-)

diff --git a/drivers/net/wireless/realtek/rtl8852bs/core/rtw_chplan.c b/drivers/net/wireless/realtek/rtl8852bs/core/rtw_chplan.c
index 55229d6eabb8..dc1a4940d3b4 100644
--- a/drivers/net/wireless/realtek/rtl8852bs/core/rtw_chplan.c
+++ b/drivers/net/wireless/realtek/rtl8852bs/core/rtw_chplan.c
@@ -3301,7 +3301,7 @@ u8 rtw_get_chplan_hdl(_adapter *adapter, u8 *pbuf)
 
 	chplan->chs_len = chset->chs_len;
 	_rtw_memcpy(chplan->chs, chset->chs, sizeof(RT_CHANNEL_INFO) * chset->chs_len);
-	*param->chplan = chplan;
+	param->chplan = chplan;
 
 	return	H2C_SUCCESS;
 }
@@ -3321,12 +3321,13 @@ u8 rtw_get_chplan_cmd(_adapter *adapter, int flags, struct get_chplan_resp **chp
 	parm = rtw_zmalloc(sizeof(*parm));
 	if (parm == NULL)
 		goto exit;
-	parm->chplan = chplan;
 
 	if (flags & RTW_CMDF_DIRECTLY) {
 		/* no need to enqueue, do the cmd hdl directly and free cmd parameter */
-		if (H2C_SUCCESS == rtw_get_chplan_hdl(adapter, (u8 *)parm))
+		if (H2C_SUCCESS == rtw_get_chplan_hdl(adapter, (u8 *)parm)) {
+			*chplan = parm->chplan;
 			res = _SUCCESS;
+		}
 		rtw_mfree((u8 *)parm, sizeof(*parm));
 	} else {
 		/* need enqueue, prepare cmd_obj and enqueue */
@@ -3363,10 +3364,11 @@ u8 rtw_get_chplan_cmd(_adapter *adapter, int flags, struct get_chplan_resp **chp
 			parm = rtw_zmalloc(sizeof(*parm));
 			if (parm == NULL)
 				goto exit;
-			parm->chplan = chplan;
 
-			if (H2C_SUCCESS == rtw_get_chplan_hdl(adapter, (u8 *)parm))
+			if (H2C_SUCCESS == rtw_get_chplan_hdl(adapter, (u8 *)parm)) {
+				*chplan = parm->chplan;
 				res = _SUCCESS;
+			}
 
 			rtw_mfree((u8 *)parm, sizeof(*parm));
 		}
diff --git a/drivers/net/wireless/realtek/rtl8852bs/core/rtw_chplan.h b/drivers/net/wireless/realtek/rtl8852bs/core/rtw_chplan.h
index ca53664ba540..be9df99fbaa0 100644
--- a/drivers/net/wireless/realtek/rtl8852bs/core/rtw_chplan.h
+++ b/drivers/net/wireless/realtek/rtl8852bs/core/rtw_chplan.h
@@ -460,7 +460,7 @@ struct get_chplan_resp {
 };
 
 struct get_channel_plan_param {
-	struct get_chplan_resp **chplan;
+	struct get_chplan_resp *chplan;
 };
 
 u8 rtw_get_chplan_hdl(_adapter *adapter, u8 *pbuf);
-- 
2.47.0

